### PR TITLE
abstract away http

### PR DESCRIPTION
**Pull request by bafolts**
I want to add support for an offline mode and allow the current server javascript to run in a worker. Currently we rely on http.IncomingMessage and http.ServerResponse for taking in a request and serving response. We don't use many parts of these types yet require the full type. This change creates a Request and Response type which picks from http for the parts we use. If nothing else this makes unit testing simpler as we only have to create the smaller Request and Response objects versus the full http.IncomingMessage and http.ServerResponse.

### CODE REVIEW

File: tests/routes/HttpMocks.ts [hunk 1]
```diff
@@ -1,29 +1,23 @@
...
     if (content) {
       this.content += content;
     }
   }
   public writeHead(statusCode: number): void {
     this.statusCode = statusCode;
   }
-
-  // MockResponse isn't really of type ServerResponse,
-  // and this makes it seem like it is so we can sneak
```
Comment (kberg) [inline tests/routes/HttpMocks.ts:24]: ooooo nice

File: tests/routes/PlayerInput.spec.ts [hunk 2]
```diff
@@ -11,16 +10,17 @@ import {RouteTestScaffolding} from './RouteTestScaffolding';
...
 
 describe('PlayerInput', function() {
   let scaffolding: RouteTestScaffolding;
-  let req: EventEmitter;
+  let req: Request;
   let res: MockResponse;
 
   beforeEach(() => {
-    req = new EventEmitter();
+    req = new EventEmitter() as unknown as Request;
```
Comment (kberg) [inline tests/routes/PlayerInput.spec.ts:21]: What happens if EventEmitter isn't forced into being Request?

File: src/server/Request.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+
+import type * as http from 'http';
+
+export type Request = Pick<http.IncomingMessage, 'headers' | 'method' | 'url'> & {
```
Comment (kberg) [inline src/server/Request.ts:4]: Can you add something close to this, please? Repeat for Response.

```
// An abstraction of http.IncomingMessage, exposing only those methods the app needs.
//
// This abstraction enables launching a serverless in-browser copy of the game.
```

**Review by kberg — APPROVED** [review]
Modulo a question and a request for some code comments, this looks great.

File: tests/routes/PlayerInput.spec.ts (hunk 2 above)
Comment (bafolts) [reply tests/routes/PlayerInput.spec.ts:21]: I can update this so it doesn't have to be forced to Request and make a real Request instance. Without this forced to Request typescript compilation fails with a compile time error.

Comment (kberg) [reply tests/routes/PlayerInput.spec.ts:21]: Let's merge this now. We can fix this later. It's for testing after all.

Comment (bafolts) [reply tests/routes/PlayerInput.spec.ts]: It made it in. Ended up being MockRequest and it fell into place pretty nicely.

https://github.com/terraforming-mars/terraforming-mars/blob/main/tests/routes/HttpMocks.ts#L5

**Pull request merged**